### PR TITLE
feat: add hide_credentials for multi-auth plugin

### LINKED ISSUE #11069: feat: Enable hide_credentials for multi-auth plugin
Current implementation supports hide_credentials in the auth_plugin sub plugin level, but in that case users can't use same header name for multiple authentication. 
example : 
`    plugins:
      multi-auth:
        auth_plugins:
          - basic-auth: { }
          - jwt-auth:
              cookie: jwt
              query: jwt
              header: authorization
          - key-auth:
              query: apikey
              header: authorization
`

### PR DESCRIPTION
**Pull request by madhawa-gunasekara**
Fixes #11069 11069

### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [x] I have explained the changes or the new features added to this PR
- [x] I have added tests corresponding to this change
- [x] I have updated the documentation to reflect this change
- [x] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

**Comment by shreemaan-abhishek** [comment]
Hi I don't understand this feature well. Could you please describe this in more detail in the linked issue?

**Comment by madhawa-gunasekara** [comment]
Hi @shreemaan-abhishek,

multi-auth plugin works as a wrapper to enable multiple authentication for routes. but I found there is a bug where users can't use a common header name for multiple authentication with credential hiding option. which we need fixed, most of the cases users need to hide their credentials. I fixed that with this pull request. this fix will allow users to use a common header to use multiple authentication with credential hiding.

**Comment by shreemaan-abhishek** [comment]
does this bug exist only in `jwt-auth` when used with `multi-auth`? I suppose other auth plugins would need fixing too.

**Comment by madhawa-gunasekara** [comment]
Hi @shreemaan-abhishek, 

most of the auth plugins use http headers, few use query params and cookies, jwt-auth and key-auth are some of them. In this fix we clear the auth-details after iterating thru all the auth plugins. Therefore I moved a generic code snippet where it clears the cookies to utils, to avoid code redundency.
 AFAIK with this fix all the auth-plugins are covered with multi-auth plugin.

File: t/plugin/multi-auth.t [hunk 1]
```diff
@@ -412,3 +412,154 @@ hello world
...
+        }
+    }
+--- request
+GET /t
+--- response_body
+passed
+
+
+
+=== TEST 15: verify key-auth with same header
```
Comment (shreemaan-abhishek) [inline t/plugin/multi-auth.t:471]: what does "same header" mean here?

File: t/plugin/multi-auth.t [hunk 2]
```diff
@@ -412,3 +412,154 @@ hello world
...
+                            ]
+                        }
+                    },
+                    "upstream": {
+                        "nodes": {
+                            "127.0.0.1:1980": 1
+                        },
+                        "type": "roundrobin"
+                    },
+                    "uri": "/hello"
```
Comment (shreemaan-abhishek) [inline t/plugin/multi-auth.t:454]: please use /echo as the uri.

/echo will return whatever request headers it found as response headers.

File: t/plugin/multi-auth.t [hunk 3]
```diff
@@ -412,3 +412,154 @@ hello world
...
+passed
+
+
+
+=== TEST 15: verify key-auth with same header
+--- request
+GET /hello
+--- more_headers
+Authorization: auth-one
+--- response_body
```
Comment (shreemaan-abhishek) [inline t/plugin/multi-auth.t:476]: please add the header check as well

**Comment by shreemaan-abhishek** [comment]
> AFAIK with this fix all the auth-plugins are covered with multi-auth plugin.

please cover them with test cases. Thank you.

File: apisix/utils/plugin-util.lua
[the remark this answers is not in the document]
Comment (madhawa-gunasekara) [reply apisix/utils/plugin-util.lua:1]: noted .

**Comment by shreemaan-abhishek** [comment]
@madhawa-gunasekara please avoid force pushing and resolve the lint failure thanks.

File: apisix/utils/auth-util.lua [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+local ngx  = ngx
+local table_insert = table.insert
+local table_concat = table.concat
+local ngx_re_gmatch = ngx.re.gmatch
+
+
+local _M = {}
+
+
```
Comment (shreemaan-abhishek) [inline apisix/utils/auth-util.lua:28]: only two blanklines are needed

File: apisix/utils/auth-util.lua [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+            break
+        end
+        if m[1] ~= key then
+            table_insert(t, m[0])
+        end
+    end
+
+    return table_concat(t, "; ")
+end
+
```
Comment (shreemaan-abhishek) [inline apisix/utils/auth-util.lua:57]: one more blank line needed

File: docs/en/latest/plugins/multi-auth.md [hunk 6]
```diff
@@ -41,6 +41,7 @@ For Route:
 | Name         | Type  | Required | Default | Description                                                           |
 |--------------|-------|----------|---------|-----------------------------------------------------------------------|
 | auth_plugins | array | True     | -       | Add supporting auth plugins configuration. expects at least 2 plugins |
+|hide_credentials|	boolean|	False|	false|	Set to true will not pass the authorization request of header\query\cookie to the Upstream.|
```
Comment (shreemaan-abhishek) [inline docs/en/latest/plugins/multi-auth.md:44]: the "`|`" looks really misplaced and you could improve the formatting a little (without changing other lines ofcourse)

**Pull request closed without merging**